rsocket/statemachine/StreamFragmentAccumulator.h

```diff
@@ -0,0 +1,69 @@
+#pragma once
+
+#include "rsocket/Payload.h"
+#include "rsocket/framing/Frame.h"
+#include "rsocket/framing/FrameHeader.h"
+
+namespace rsocket {
+
+struct StreamFragmentAccumulator {
+  FrameHeader header;
+  uint32_t requestN{0};
+  Payload fragments;
+
+  StreamFragmentAccumulator(StreamFragmentAccumulator const&) = delete;
+  StreamFragmentAccumulator(StreamFragmentAccumulator&&) = default;
+
+  StreamFragmentAccumulator() {}
+
+  explicit StreamFragmentAccumulator(
+      Frame_REQUEST_Base const& frame,
+      Payload payload)
+      : StreamFragmentAccumulator(
+            frame.header_,
+            frame.requestN_,
+            std::move(payload)) {}
+
+  explicit StreamFragmentAccumulator(
+      Frame_REQUEST_RESPONSE const& frame,
+      Payload payload)
+      : StreamFragmentAccumulator(frame.header_, 0, std::move(payload)) {}
+
+ private:
+  explicit StreamFragmentAccumulator(
+      FrameHeader const& fh,
+      uint32_t requestN,
+      Payload payload)
+      : header(fh), requestN(requestN) {
+    addPayload(std::move(payload));
+  }
+
+ public:
+  void addPayload(Payload p) {
+    if (p.metadata) {
+      if (!fragments.metadata) {
+        fragments.metadata = std::move(p.metadata);
+      } else {
+        fragments.metadata->prev()->appendChain(std::move(p.metadata));
+      }
+    }
```

Comment:
would you dislike using folly::IOBufQueue for metadata and data to avoid this if statement and messing up with iobuf internal structure?

Reply:
It'd mean making `StreamFragmentAccumulator` quite a bit heavier. Also we need to track if the final `fragments.metadata` should be a nullptr or not; an IOBufQueue would unconditionally return an empty non-null `IOBuf` pointer when moving its accumulator. More complex than it needs to be.